File: Problem0006/Problem6.py
# -*- coding: utf-8 -*-
"""
The sum of the squares of the first ten natural numbers is,
    1**2 + 2**2 + ... + 10**2 = 385
The square of the sum of the first ten natural numbers is,
    (1 + 2 + ... + 10)**2 = 55**2 = 3025
Hence the difference between the sum of the squares of the first ten natural
numbers and the square of the sum is 3025 − 385 = 2640.
Find the difference between the sum of the squares of the first one hundred
natural numbers and the square of the sum.

Answer:
    25164150
Completed on Tue, 1 Apr 2014, 17:57
"""


def main(n=100):
    natural_numbers = range(1, n+1)
    square_sum = sum(natural_numbers) ** 2
    sum_squares = sum(n ** 2 for n in natural_numbers)
    return square_sum - sum_squares

# # # #


if __name__ == '__main__':
    print main()

File: Problem0006/Problem6.hs
-- Find the difference 
-- between the sum of the squares of the first one hundred natural numbers 
-- and the square of the sum.

-- Solving this in Haskell is almost trivial
sumPower n e = sum [x^e | x <- [1..n]]
powerSum n e = (sum [1..n])^e
-- This returns in under a minute all the way up to
-- 10000000

-- But we are still doing a lot of work we don't need to.
-- First, sum [1..n] is a simple geometric series 
-- which has a well known formula.
-- This allows us to greatly improve our powerSum
powerSum' n e = ((n*(n+1))/2)^e
-- Second, sum [x^e | x <- [1..n]] is a simple power series.
-- Unfortunately this power series does not have a formula
-- for the generic case, so we'll have to focus on the 
-- specific case of sum [x^2 | x <- [1..n]]
sumSquares n = ((n^3)/3) + ((n^2)/2) + (n/6)

-- (powerSum' 100 2) - (sumSquares 100)
-- 25164150